docs/lifecycle.md

```diff
@@ -0,0 +1,278 @@
+  * if any action is enabled, create the zookeeper node for this
+    owner/bucket (ignore if already exists)
+  * if no action is enabled, delete the zookeeper node for this
+    owner/bucket (ignore if not found)
+* when processing a log entry that is a deletion of a bucket, delete
+  the zookeeper node for that owner/bucket (ignore if not found).
+
+The resulting zookeeper list should contain a current view of all
+buckets that have lifecycle enabled, up to the point where the
+metadata log has been processed.
+
+### Conductor
+
+The conductor role is to periodically do the following (cron-style task):
+
+* create a zookeeper node at `/[chroot_path]/lifecycle/run/conductorLock`
+  with EPHEMERAL flag (so that it gets cleaned up automatically in
+  case the zookeeper client session expires)
+  * if ZNODEEXISTS error is returned, it's already being processed by
+    another conductor instance, so do nothing until the next cron
+    iteration
+* get the list of buckets with potential lifecycle actions enabled
+  from zookeeper (getChildren in zookeeper API)
+* for each of them, do the following:
+  * create a zookeeper node at
+    `/[chroot_path]/lifecycle/run/queuedBuckets/${ownerId}:${bucket}`
+    with EPHEMERAL flag, i.e. mark the bucket as being queued
+    * If ZNODEEXISTS error is returned, it's already queued, so stop
+      and go on with the next bucket
+  * publish a message to the kafka topic
+    *backbeat-lifecycle-bucket-tasks* (see [bucket tasks](#bucket-tasks)
+    message format)
+* remove the `conductorLock` node
+
+The conductor will only generate messages to *start* a new
+listing. Later on, if the listing is truncated, new messages will be
+generated by the Producer to create new tasks to continue the
+listing. This way, we can efficiently parallelize the processing of
+large buckets across producers (see [Producer](#producer) section).
+
+Note that as an optimization, when lifecycle rules only apply to a
+particular object prefix in a bucket, it can be interesting to list
+only the specific range belonging to this prefix. For this we can add
+the necessary info in the message JSON data to limit the listing to
+this range.
+
+### Producer
+
+The producer's responsibility is to list objects in lifecycled
+buckets, and publish to kafka queues the actual lifecycle actions to
+execute on objects. For now, supported actions are expiration actions.
+
+More specifically, the producer is part of a kafka consumer group that
+processes items from the *backbeat-lifecycle-bucket-tasks* topic. For
+each of them, it does the following:
+
+* Generate or use cached credentials to be able to query the target
+  bucket (through admin-backbeat.json?)
+* Get the lifecycle configuration from S3 connector of the target
+  bucket (GET /foobucket?lifecycle)
+  * If no lifecycle configuration is associated to the bucket, skip
+    the rest of bucket processing
+* List the object versions in the bucket, starting from the
```

Comment:
Just to clarify the steps here:
- list every object in the bucket (which defaults to a max of 1000 listings)
- If exceeds 1000 object listings, re-queue to Kafka w/ `StartAfter` to get next object listings
- For each object, if versioned, list the objects versions (defaults to a max of 1000 listings again).
- If versioned and exceeds 1000 versions, need to re-queue to Kafka w/ `KeyMarker` and `VersionIdMarker`.
- If not versioned, just go ahead and queue up the delete request for the object to the consumer. If versioned, queue up the delete request to the consumer for every version of the object.

Pre-filter step needed to determine if the Kafka entry is a bucket or a versioned object w/ marker. I'm not sure how difficult this step will be at this moment, but the "Bucket tasks" Kafka entry might need to have some sort of indicator. Maybe just based on if the entry holds properties like `KeyMarker` or `StartAfter`?